src/data_types/tensor.hpp

```diff
@@ -0,0 +1,410 @@
+
+private:
+    static constexpr std::size_t s_n_elements = (ddc::type_seq_size_v<ValidIndexSet> * ...);
+    std::array<ElementType, s_n_elements> m_data;
+
+public:
+    /// The type of the elements of the tensor.
+    using element_type = ElementType;
+
+    /**
+     * @brief The rank of the tensor.
+     * This is equivalent to the number of indices required to
+     * access an element of the tensor.
+     *
+     * @return The rank of the tensor.
+     */
+    KOKKOS_FUNCTION static constexpr std::size_t rank()
+    {
+        return sizeof...(ValidIndexSet);
+    }
+
+    /**
+     * @brief The size of the tensor.
+     * This is the number of elements in the tensor.
+     *
+     * @return The number of elements in the tensor.
+     */
+    KOKKOS_FUNCTION static constexpr std::size_t size()
+    {
+        return s_n_elements;
+    }
+
+private:
+    static_assert(rank() > 0);
+
+public:
+    /**
+     * @brief Construct an uninitialised tensor object.
+     */
+    explicit KOKKOS_FUNCTION Tensor() {}
+
+    /**
+     * @brief Construct a tensor object initialised with a value.
+     * @param fill_value The value with which the tensor should be filled.
+     */
+    explicit KOKKOS_FUNCTION Tensor(ElementType fill_value)
+    {
+        m_data.fill(fill_value);
+    }
+
+    /**
+     * @brief Construct a 1D tensor object by providing the elements that should
+     * be saved in the vector.
+     *
+     * @param elements The elements of the tensor.
+     */
+    template <
+            class... Params,
+            class = std::enable_if_t<(std::is_convertible_v<Params, ElementType> && ...)>,
+            class = std::enable_if_t<sizeof...(Params) == size()>>
+    KOKKOS_FUNCTION Tensor(Params... elements) : m_data({elements...})
+    {
+        static_assert(
+                rank() == 1,
+                "Filling the tensor on initialisation is only permitted for 1D vector objects");
+    }
+
+    /**
+     * @brief Construct a tensor object by copying an existing tensor.
+     *
+     * @param o_tensor The tensor to be copied.
+     */
+    template <class OElementType>
+    explicit KOKKOS_FUNCTION Tensor(Tensor<OElementType, ValidIndexSet...> o_tensor)
+        : m_data(o_tensor.m_data)
+    {
+    }
+
+    /**
+     * @brief Get a modifiable reference to an element of the tensor.
+     * @tparam QueryIndexTag A type describing the relevant index.
+     * @return The relevant element of the tensor.
+     */
+    template <class QueryTensorIndexElement>
+    KOKKOS_FUNCTION ElementType& get()
+    {
+        static_assert(tensor_tools::is_tensor_index_element_v<QueryTensorIndexElement>);
+        return m_data[QueryTensorIndexElement::index()];
+    }
+
+    /**
+     * @brief Get an element of the tensor.
+     * @tparam QueryIndexTag A type describing the relevant index.
+     * @return The relevant element of the tensor.
+     */
+    template <class QueryTensorIndexElement>
+    KOKKOS_FUNCTION ElementType get() const
```

Comment:
```suggestion
    KOKKOS_FUNCTION ElementType const& get() const
```
We usually return a `const&`, like `std::array` does.

Reply:
Fixed 7efec89693646ff8d17eff1d9fccc363f9f5e2a5